plugins/search-backend-node/src/engines/LunrSearchEngine.ts

```diff
@@ -50,40 +50,62 @@ export class LunrSearchEngine implements SearchEngine {
     filters,
     types,
   }: SearchQuery): ConcreteLunrQuery => {
-    let lunrQueryFilters;
-    const lunrTerm = term ? `+${term}` : '';
-    if (filters) {
-      lunrQueryFilters = Object.entries(filters)
-        .map(([field, value]) => {
-          // Require that the given field has the given value (with +).
-          if (['string', 'number', 'boolean'].includes(typeof value)) {
-            if (typeof value === 'string') {
-              return ` +${field}:${value.replace(':', '\\:')}`;
-            }
-            return ` +${field}:${value}`;
-          }
-
-          // Illustrate how multi-value filters could work.
-          if (Array.isArray(value)) {
-            // But warn that Lurn supports this poorly.
-            this.logger.warn(
-              `Non-scalar filter value used for field ${field}. Consider using a different Search Engine for better results.`,
-            );
-            return ` ${value.map(v => {
-              return `${field}:${v}`;
-            })}`;
-          }
+    return {
+      lunrQueryBuilder: q => {
+        const termToken = lunr.tokenizer(term);
+
+        // Support for typeahead seach is based on https://github.com/olivernn/lunr.js/issues/256#issuecomment-295407852
+        // look for an exact match and apply a large positive boost
+        q.term(termToken, {
```

Comment:
This is really cool, but it makes me wonder if we should more clearly delineate what logic should go in a `*SearchEngine`'s default translator vs. what should go elsewhere.  Especially starting to see specific values in boost fields and things like that.  This gets even fuzzier when we start to consider custom translation logic that depends on the context of the search (e.g. you may want to tweak boost values and things differently when searching across _all_ TechDocs sites, vs. when you're just looking at _one_ TechDocs site).

That being said, having this readily available as _a_ translator that could be easily plugged in (and iterated on by the community) seems valuable.

What do you think about separating out the fuzzy-logic into like a `LunrTranslatorFuzzy` that could be swapped in by integrators like...

```typescript
import {
  LunrSearchEngine,
  LunrTranslatorFuzzy,
} from '@backstage/plugin-search-backend-node';

const searchEngine = new LunrSearchEngine({ logger });
searchEngine.setTranslator(LunrTranslatorFuzzy);
```

And in the default translator, keeping a simpler...

```typescript
q.term(termToken, {
  usePipeline: true,
  presence: lunr.Query.presence.REQUIRED,
})
```

Reply:
Could even break out and export a `LunrTranslatorDefault` that could be used in that context-based conditional logic by app integrators too?

```typescript
import {
  LunrSearchEngine,
  LunrTranslatorDefault,
  LunrTranslatorFuzzy,
} from '@backstage/plugin-search-backend-node';
import { AcmeCorpCustomLunrTranslator } from './translators'

const searchEngine = new LunrSearchEngine({ logger });
searchEngine.setTranslator(query => {
  if (query.types[0] === '*') {
    return LunrTranslatorFuzzy(query);
  }
  if (someOtherSpecificCondition) {
    return AcmeCorpCustomLunrTranslator(query);
  }
  return LunrTranslatorDefault(query);
});
```

Reply:
If you think the default logic is something we want to keep (especially with `presence: lunr.Query.presence.REQUIRED`)   then I'm open to provide it as a second transformer as you suggested 👍

Reply:
Yeah, I'm a little on the fence on it.  It comes down to what we want a default translator to represent.

**If a default translator should just be an MVP** (e.g. the bare minimum code necessary to wire up each part of the search query to its representation in the search engine's query language), then we'll have a fairly objective way to evaluate changes, at the expense of the results being bad in ~80% of cases.  ...Consequences could be: more issue reports about bad search results and/or it basically becoming a requirement that people write/set custom translators.

**If a default translator should be "good enough for 80%"** (e.g. something like what's in this PR), folks will usually have a good experience out of the box, but it becomes really difficult to evaluate iterations on the translator ("does this change make things better or worse?").  Beyond review uncertainty, another consequence would be that folks might be less likely to set up custom translators, increasing the risk that a change to the default translator would negatively affect someone's search setup out in the wild.

This change _feels_ really good, but when you start poking at the edges, it's harder to be objective.  e.g. Try searching for `noresults` with the `kind` filter set to `Component` in the default branch vs. this one.  The default branch returns no results, while this branch returns every Component entity, ranked by how fuzzily they match `noresults`.  Which one's better?  🤷‍♂️   It depends on your perspective...

---

Starting to build up a library of contributed translators, while keeping the default one a little naive, might be the best of both worlds...  The default one makes search work _at all_, but then integrators would have a smörgåsbord of easy-to-import translators that they could use if they wanted.  Plus they'd have lots of examples on how to write their own.

...Or maybe that's overkill for Lunr and something we should only think about for "real" search engines like Elastic or Postgres?

I dunno.  Any thoughts @emmaindal or @soapraj?

Reply:
After sleeping on it: I think this is fine as-is.  We can get a little more strategic about this in "production" search-engines.